### PR TITLE
Refactor of AMP Ads API into handler

### PR DESCRIPTION
**Pull request by tdrl**
This separates out (what will be) shared functionality from the amp-ad-3p-impl module into a standalone.  In a future commit on this branch, it will be integrated with amp-a4a, so that it will share API behavior with 3p-impl.

NOT YET TESTED.

### CODE REVIEW

**Comment by keithwrightbos** [comment]
I have reworked the code to remove hard dependencies between 3p and the new API handler as well as wired into A4A.  I have to fix a few test failures within amp-ad otherwise I think this is good to go.  Let me know what you think.  The reviewer has become the reviewee!

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+  }
+
+  /**
+   * @param {!Element} iframe
+   * @param {boolean} is3p
+   * @return {!Promise} awaiting load event for ad frame
+   */
+  startUp(iframe, is3p) {
+    user.assert(
+      !this.iframe, 'multiple invocations of startup without destroy!');
```
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: this.iframe_ ?

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+      } else {
+        user.info('no content callback was specified');
+      }
+    }, this.is3p_);
+    // Triggered by context.reportRenderedEntityIdentifier(…) inside the ad
+    // iframe.
+    listenForOnce(this.iframe_, 'entity-id', info => {
+      this.element_.creativeId = info.id;
+    }, this.is3p_);
+    this.unlisteners_.push(listenFor(this.iframe_, 'embed-size', data => {
```
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Does this need to be a list?  I could only find one listenFor occurrence, and if we can't have multiple in flight at once, do we need to keep a list of unlisteners?

File: extensions/amp-ad/0.1/amp-ad-3p-impl.js [hunk 3]
```diff
@@ -159,6 +159,9 @@ export class AmpAd3PImpl extends AMP.BaseElement {
     /** @private {?Element} */
     this.iframe_ = null;
```
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad-3p-impl.js]: Should this class still have an iframe_ parameter?  I'd expect  the ApiHandler to exclusively own the iframe.

File: extensions/amp-ad/0.1/amp-ad-api-handler.js [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+    this.is3p_ = false;
+
+    /** @param {?function} opt_noContentHandler */
+    this.noContentCallback_ = opt_noContentCallback;
+
+    /** @private {!Array<!Function>} functions to unregister listeners */
+    this.unlisteners_ = [];
+
+    /** @private @const */
+    this.viewer_ = viewerFor(this.baseInstance_.getWin());
```
Comment (tdrl) [inline extensions/amp-ad/0.1/amp-ad-api-handler.js]: Yeah, so this is one of the points I was unsure about that created some of the complication.  Can the viewer live at the ApiHandler level, or should it really live at the BaseElement instance level?

**Comment by tdrl** [comment]
Overall structure looks good -- much clearer than before.  I see that the solution was delegating the callbacks through -- what had hung me up, really, is that ApiHandler isn't a BaseElement, so won't receive all the AMP callbacks itself.  But if you delegate them through, then it's much simpler.

If the tests pass, I'm happy.

**Pull request merged**